File: fabric/src/datagen/java/dev/mayaqq/estrogen/datagen/impl/loottables/EstrogenLootTables.java
package dev.mayaqq.estrogen.datagen.impl.loottables;

import dev.mayaqq.estrogen.registry.EstrogenBlocks;
import dev.mayaqq.estrogen.registry.blocks.ModelBedBlock;
import net.fabricmc.fabric.api.datagen.v1.FabricDataOutput;
import net.fabricmc.fabric.api.datagen.v1.provider.FabricBlockLootTableProvider;
import net.minecraft.world.level.block.DoorBlock;
import net.minecraft.world.level.block.state.properties.BedPart;
import net.minecraft.world.level.block.state.properties.DoubleBlockHalf;

public class EstrogenLootTables extends FabricBlockLootTableProvider {

    public EstrogenLootTables(FabricDataOutput dataOutput) {
        super(dataOutput);
    }

    @Override
    public void generate() {
        add(EstrogenBlocks.CENTRIFUGE.get(), createSingleItemTable(EstrogenBlocks.CENTRIFUGE.get()));
        add(EstrogenBlocks.COOKIE_JAR.get(), createSilkTouchOnlyTable(EstrogenBlocks.COOKIE_JAR.get()));
        add(EstrogenBlocks.DORMANT_DREAM_BLOCK.get(), createSingleItemTable(EstrogenBlocks.DORMANT_DREAM_BLOCK.get()));
        add(EstrogenBlocks.ESTROGEN_PILL_BLOCK.get(), createSingleItemTable(EstrogenBlocks.ESTROGEN_PILL_BLOCK.get()));
        add(EstrogenBlocks.MOTH_WOOL.get(), createSingleItemTable(EstrogenBlocks.MOTH_WOOL.get()));
        add(EstrogenBlocks.MOTH_SEAT.get(), createSingleItemTable(EstrogenBlocks.MOTH_SEAT.get()));
        add(EstrogenBlocks.QUILTED_MOTH_WOOL.get(), createSingleItemTable(EstrogenBlocks.QUILTED_MOTH_WOOL.get()));
        add(EstrogenBlocks.QUILTED_MOTH_WOOL_CARPET.get(), createSingleItemTable(EstrogenBlocks.QUILTED_MOTH_WOOL_CARPET.get()));
        add(EstrogenBlocks.MOTH_WOOL_CARPET.get(), createSingleItemTable(EstrogenBlocks.MOTH_WOOL_CARPET.get()));
        add(EstrogenBlocks.MOTH_BED.get(), createSinglePropConditionTable(EstrogenBlocks.MOTH_BED.get(), ModelBedBlock.PART, BedPart.HEAD));
        add(EstrogenBlocks.QUILTED_MOTH_BED.get(), createSinglePropConditionTable(EstrogenBlocks.QUILTED_MOTH_BED.get(), ModelBedBlock.PART, BedPart.HEAD));
    }
}

File: src/fabric/data/kotlin/dev/mayaqq/estrogen/datagen/impl/loottables/EstrogenLoottables.kt
package dev.mayaqq.estrogen.datagen.impl.loottables

import dev.mayaqq.estrogen.content.EstrogenBlocks
import dev.mayaqq.estrogen.content.EstrogenComponents
import net.fabricmc.fabric.api.datagen.v1.FabricDataOutput
import net.fabricmc.fabric.api.datagen.v1.provider.FabricBlockLootTableProvider
import net.minecraft.core.HolderLookup
import net.minecraft.world.level.block.BedBlock
import net.minecraft.world.level.block.state.properties.BedPart
import net.minecraft.world.level.storage.loot.functions.CopyComponentsFunction
import java.util.concurrent.CompletableFuture


class EstrogenLoottables(output: FabricDataOutput, lookup: CompletableFuture<HolderLookup.Provider>) :
    FabricBlockLootTableProvider(output, lookup)
{

    override fun generate() {
        add(EstrogenBlocks.CookieJar.value!!, createSilkTouchOnlyTable(EstrogenBlocks.CookieJar.value!!))
        add(EstrogenBlocks.DreamBlock.get(), createSingleItemTable(EstrogenBlocks.DreamBlock.get()))
        add(EstrogenBlocks.EstrogenPillBlock.value!!, createSingleItemTable(EstrogenBlocks.EstrogenPillBlock.value!!))
        add(EstrogenBlocks.MothWool.value!!, createSingleItemTable(EstrogenBlocks.MothWool.value!!))
        add(EstrogenBlocks.QuiltedMothWool.value!!, createSingleItemTable(EstrogenBlocks.QuiltedMothWool.value!!))
        add(EstrogenBlocks.QuiltedMothCarpet.value!!, createSingleItemTable(EstrogenBlocks.QuiltedMothCarpet.value!!))
        add(EstrogenBlocks.MothCarpet.value!!, createSingleItemTable(EstrogenBlocks.MothCarpet.value!!))
        add(
            EstrogenBlocks.MothBed.value!!,
            createSinglePropConditionTable(EstrogenBlocks.MothBed.value!!, BedBlock.PART, BedPart.HEAD)
        )
        add(
            EstrogenBlocks.QuiltedMothBed.value!!,
            createSinglePropConditionTable(EstrogenBlocks.QuiltedMothBed.value!!, BedBlock.PART, BedPart.HEAD)
        )
        add(EstrogenBlocks.DreamCatcher.value!!, createSingleItemTable(EstrogenBlocks.DreamCatcher.value!!).apply(
            CopyComponentsFunction.copyComponents(CopyComponentsFunction.Source.BLOCK_ENTITY).include(
                EstrogenComponents.TriColorComponent
            )
        ))
    }
}